### DISCUSSION TITLE
Compiling C firmware: how to insert version number into header file BEFORE building the firmware?

### DISCUSSION THREAD

**Discussion by coaxial** [semantic-release/semantic-release#2860] [Release workflow]
I'm trying to use semantic-release with a C firmware project.

I'd like to show the current version on the device, which is fetched from a header file with a `#define`.

In `.releaserc.json`, I have:

```json
{
  "plugins": [
        [
      "@semantic-release/exec",
      {
        "prepareCmd": "./scripts/version_date.sh ${nextRelease.version} `date \"+%Y-%m-%d\"`"
      }
    ],
  ]
}
```

`version_date.sh` echoes the right values into the header file.

My problem is that I have to build the firmware first, and then semantic-release runs. As such, the version header file is only updated after the firmware has been built and the version in the compiled artifact is wrong.

How could I solve it? I'm using GitHub Actions (especially to build the firmware), and the workflows are here: https://github.com/coaxial/qmk_keymap/tree/e4c950afa212b19ce3841f1c1666e23388a60a62/.github/workflows. I don't think there is a way to run another GitHub Action as part `@semantic-release/exec`'s `prepareCmd`?

**Comment by gr2m** [comment]
Can you build the firmware as part of semantic release using `@semantic-release/exec`? You can use the same plugin multiple times with different options in the `plugins` config

**Reply by coaxial** [reply]
The build process is not straightforward and uses an action. I don't believe there is a way to trigger a GitHub action in the middle of semantic-release running, is there?

You gave me an idea though. Maybe I can have two config files with the first one only calculating the version number, committing it to a file, then running the compile workflow, then running the actual semantic-release flow to create the release.

**Reply by gr2m** [reply]
> The build process is not straightforward and uses an action. I don't believe there is a way to trigger a GitHub action in the middle of semantic-release running, is there?

no

> You gave me an idea though. Maybe I can have two config files with the first one only calculating the version number, committing it to a file, then running the compile workflow, then running the actual semantic-release flow to create the release.

It might be tricky to get the version without semantic-release also creating the tag, which won't make it work the same when you run it a 2nd time. It's tricky, there is a long standing open issue about only retrieving the version without anything else, it would help here in your use case as well. 

I hope you find a way!

**Comment by seebeen** [comment]
> It might be tricky to get the version without semantic-release also creating the tag, which won't make it work the same when you run it a 2nd time.

This is the main reason I made the plugin I mentioned in another discussion. Since the version tag is determined by commits. Is that something that could be done via a separate CLI command which would just invoke the commit analyzer and spit out the next tag?
